File: tests/fixtures/sample.swift
import Foundation

protocol PaymentProcessor {
    func process(amount: Int64) throws -> TransactionId
}

// Implements PaymentProcessor on top of the StripeClient SDK.
final class StripePaymentProcessor: PaymentProcessor {
    private let apiKey: String

    init(apiKey: String) {
        self.apiKey = apiKey
    }

    func process(amount: Int64) throws -> TransactionId {
        print("charging via StripeClient")
        return try charge(amount: amount)
    }

    private func charge(amount: Int64) throws -> TransactionId {
        return TransactionId.of(amount)
    }
}

struct TransactionId {
    let raw: String

    static func of(_ amount: Int64) -> TransactionId {
        return TransactionId(raw: "tx-\(amount)")
    }
}

File: tests/fixtures/sample.lua
-- Payment processing on top of the StripeClient binding.
local TransactionId = require("billing.transaction_id")

local StripePaymentProcessor = {}
StripePaymentProcessor.__index = StripePaymentProcessor

function StripePaymentProcessor.new(api_key)
    local self = setmetatable({}, StripePaymentProcessor)
    self.api_key = api_key
    return self
end

function StripePaymentProcessor:charge(amount)
    return TransactionId.of(amount)
end

function StripePaymentProcessor:process(amount)
    print("charging via StripeClient")
    return self:charge(amount)
end

local function process_batch(processor, invoices)
    for _, invoice in ipairs(invoices) do
        processor:process(invoice.amount)
    end
end

return {
    StripePaymentProcessor = StripePaymentProcessor,
    process_batch = process_batch,
}
